### ISSUE TITLE
S2583 FP: confusing message for `Guid` comparison - should trigger S2589 instead

### ISSUE THREAD

**Issue by tnoiretisa** [SonarSource/sonar-dotnet#2411]
### Description

False positive with || operator, rule add violation on first part and dont care about the other parts

### Repro steps

With this code, rule throw "Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed." on "guid == null"

        public void DoTest(Guid guid)
        {
            if (guid == null || guid == Guid.Empty)
                guid = Guid.NewGuid();
        }

### Expected behavior

Take care about others parts

### Related information

* SonarC# Version 7.11
* Visual Studio Version vs 2017

**Comment by Corniel** [comment]
This is not a false positive: `guid == null` will always return false. So this code should be changed to:

``` C#
public void DoTest(Guid guid)
    {
        if (guid == Guid.Empty)
            guid = Guid.NewGuid();
    }
```

You could argue though, that is should have a Fix to take out those parts that do not add anything to the condition.

**Comment by tnoiret** [comment]
For me it's false positive because "some subsequent code is never executed." is false, 
But i'm agree with you, best way it to remove "guid == null"

**Comment by Corniel** [comment]
I disagree: It is not a false positive. It is a a confusing message. And I agree that the message should be improved.

**Comment by andrei-epure-sonarsource** [comment]
Thank you @tnoiret and @Corniel for your contributions.

Indeed, the message is confusing and I added [a comment to RSPEC-2583](https://jira.sonarsource.com/browse/RSPEC-2583?focusedCommentId=212615&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-212615) such that we can improve it in the future.

And this seems to be rather an occurrence of [RSPEC-2589](https://rules.sonarsource.com/csharp/RSPEC-2589) _Boolean expressions should not be gratuitous_

**Comment by nicolas-harraudeau-sonarsource** [comment]
@andrei-epure-sonarsource this issue should indeed be raised by RSPEC-2589 instead of RSPEC-2583. There is no dead code in this case. It is just a maintainability issue due to an unnecessary expression.

If `guid == null` was always true then it would be correct to raise an issue with RSPEC-2583 as the second expression would never be evaluated.

**Comment by mary-georgiou-sonarsource** [comment]
Fixed in https://github.com/SonarSource/sonar-dotnet/pull/7750
